packages/@eventual/core-runtime/src/workflow/workflow-executor.ts

```diff
@@ -206,13 +207,17 @@ export class WorkflowExecutor<Input, Output, Context = undefined> {
     ) => void;
   } = {};
 
+  // clone the history so it cannot be modified by the user
+  public historyCloned: HistoryStateEvent[];
+
   constructor(
     private workflow: Workflow<any, Input, Output>,
     public history: HistoryStateEvent[],
     // TODO: properties used in the workflow should be encoded in the history to keep them constant between runs
     private propertyRetriever: PropertyRetriever,
     private eventualFactory: EventualFactory = createDefaultEventualFactory()
   ) {
+    this.historyCloned = deepClone(history);
```

Comment:
maybe use `.historyCloned` in the iterator in here, leave it private, and then leave the original array in the `.history`?